File: leetcode/combination_sum.js
// https://leetcode.com/problems/combination-sum/
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */

var combinationSum = function(candidates, target) {
    let total = [];
    recursive([],0);
 
    function push(arr) {
        total.push(arr.slice())
    }
    function recursive(curr,index) {
        if (sum(curr) > target) {
            return null
        }
        else if (sum(curr) == target) {
            return curr
        }
        for (let i = index; i < candidates.length ; i++) {
            curr.push(candidates[i]);
            let res = recursive(curr,i);
            if (res) {
                push(res);
            }
            curr.pop()
        }
        return null
    }
    return total
};


function sum(arr) {
    return arr.reduce((a,num) => a+num ,0)
}
function main() {
    let candidates =[2,3,5,6];
    let target = 8;
    let res = combinationSum(candidates,target)
    console.log(res);

}

main()

File: leetcode/combination_sum.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []

        def findSum(arr: List[int], sum: int, currIndex: int):
            if sum == target:
                result.append(arr)
                return

            if sum > target:
                return

            for index in range(currIndex, (len(candidates))):
                num = candidates[index]
                a = arr[:]
                a.append(num)
                findSum(a, sum + num, index)

        findSum([], 0, 0)
        return result


if __name__ == '__main__':
    candidates = [5, 2, 4, 3]
    target = 8
    res = Solution().combinationSum(candidates, target)
    print(res)
